newIDE/app/src/Utils/GDevelopServices/User.js

```diff
@@ -48,3 +75,68 @@ export const getUserPublicProfile = (
     .get(`${GDevelopUserApi.baseUrl}/user-public-profile/${id}`)
     .then(response => response.data);
 };
+
+const postBadgeIfLocked = (
+  authenticatedUser: AuthenticatedUser,
+  achievementId: string
+): ?Promise<?Badge> => {
+  const {
+    badges,
+    firebaseUser,
+    getAuthorizationHeader,
+    onBadgesChanged,
+  } = authenticatedUser;
+  if (!badges || !firebaseUser) return null;
+  if (badges.map(badge => badge.achievementId).includes(achievementId)) {
+    return null;
+  }
+  console.log(`posting achievement ${achievementId}`);
+
+  const userId = firebaseUser.uid;
+  return getAuthorizationHeader().then(authorizationHeader =>
+    axios
+      .post(
+        `${GDevelopUserApi.baseUrl}/user/${userId}/badge`,
+        {
+          achievementId,
+        },
+        {
+          params: {
+            userId,
+          },
+          headers: {
+            Authorization: authorizationHeader,
+          },
+        }
+      )
+      .then(response => {
+        onBadgesChanged();
+        return response.data;
+      })
+      .catch(err => {
+        if (err.response.status === 409) {
+          console.warn('Badge already exists');
+        } else {
+          throw err;
+        }
+      })
+  );
+};
+
+export const addPostBadgePreHookIfLocked = (
```

Comment:
same here, this function is hard to understand from its name (I don't have a better idea yet...)